File: examples/nirfmx/wcdma/csharp/VS2010/RFmxWcdmaAcpMultiCarrier/cs/RFmxWcdmaAcpMultiCarrier.cs
//Steps:
//1. Open a new RFmx Session.
//2. Configure Frequency Reference.
//3. Configure basic signal properties (Center Frequency, RF Attenuation and External Attenuation)
//4. Configure Trigger Type and Trigger Parameters.
//5. Configure Contiguous Carriers.
//6. Configure Reference Level.
//7. Select ACP measurement and enable Traces.
//8. Configure Measurement Method.
//9. Configure Averaging Parameters for ACP measurement.
//10. Configure Sweep Time Parameters.
//11. Configure Noise Compensation Parameter.
//12. Configure Number of offset channels. 
//13. Configure Offset Power Reference
//14. Initiate the Measurement.
//15. Fetch ACP Measurements and Traces.
//16. Close RFmx Session.
using System;
using NationalInstruments.RFmx.InstrMX;
using NationalInstruments.RFmx.WcdmaMX;

namespace NationalInstruments.Examples.RFmxWcdmaAcpMultiCarrier
{
    public class RFmxWcdmaAcpMultiCarrier
    {
        RFmxInstrMX instrSession;
        RFmxWcdmaMX wcdma;

        string resourceName = "RFSA";
        RFmxWcdmaMXMeasurementTypes measurement = RFmxWcdmaMXMeasurementTypes.Acp;
        string frequencyReferenceSource = RFmxInstrMXConstants.OnboardClock;
        double frequencyReferenceFrequency = 10.0e+6;                     /* Hz */
        double centerFrequency = 1.95e+9;               /* Hz */
        double externalAttenuation = 0.000000;          /* dB */
        RFmxInstrMXRFAttenuationAuto rfAttenuationAuto = RFmxInstrMXRFAttenuationAuto.True;
        double rfAttenuation = 10.0;                    
        string digitalEdgeSource = RFmxWcdmaMXConstants.Pfi0;
        RFmxWcdmaMXDigitalEdgeTriggerEdge digitalEdge = RFmxWcdmaMXDigitalEdgeTriggerEdge.Rising;
        double triggerDelay = 0.000000;                 /* seconds */
        double measurementInterval = 0.010000;          /* seconds */
        double referenceLevel = 0.000000;               /* dBm */
        RFmxWcdmaMXAcpMeasurementMethod measurementMethod = RFmxWcdmaMXAcpMeasurementMethod.Normal;
        RFmxWcdmaMXAcpAveragingEnabled averagingEnabled = RFmxWcdmaMXAcpAveragingEnabled.False;
        int averagingCount = 10;
        RFmxWcdmaMXAcpAveragingType averagingType = RFmxWcdmaMXAcpAveragingType.Rms;
        bool autoLevel = true;
        RFmxWcdmaMXAcpSweepTimeAuto sweepTimeAuto = RFmxWcdmaMXAcpSweepTimeAuto.True;
        double sweepTimeInterval = 6.67e-4;             /* seconds */
        RFmxWcdmaMXAcpNoiseCompensationEnabled noiseCompensationEnabled = RFmxWcdmaMXAcpNoiseCompensationEnabled.False;
        int numberOfOffsets = 2;
        double timeout = 10.000000;                     /* seconds */
        Spectrum<float> spectrum;

        bool enableAllTraces = true;
        bool enableTrigger = false;

        int numberOfCarriers = 2;
        int carrierAtCenterFrequency = -1;
        RFmxWcdmaMXAcpOffsetPowerReferenceCarrier offsetPowerReferenceCarrier = RFmxWcdmaMXAcpOffsetPowerReferenceCarrier.Composite;
        int offsetPowerReferenceSpecific = 0;
        double totalCarrierPower = 0;

        double[] lowerAbsolutePower, upperAbsolutePower, lowerRelativePower, upperRelativePower;
        double[] absolutePower, relativePower;

        public void Run()
        {
            try
            {
                InitializeInstr();
                ConfigureWcdma();
                RetrieveResults();
                PrintResults();
            }
            catch (Exception ex)
            {
                DisplayError(ex);
            }
            finally
            {
                CloseSession();
                Console.WriteLine("Press any key to exit");
                Console.ReadKey();
            }
        }

        private void InitializeInstr()
        {
            /* Create a new RFmx Session */
            instrSession = new RFmxInstrMX(resourceName, "");
        }

        private void ConfigureWcdma()
        {
            wcdma = instrSession.GetWcdmaSignalConfiguration();
            instrSession.ConfigureFrequencyReference("", frequencyReferenceSource, frequencyReferenceFrequency);
            wcdma.ConfigureFrequency("", centerFrequency);
            wcdma.ConfigureExternalAttenuation("", externalAttenuation);

            instrSession.ConfigureRFAttenuation("", rfAttenuationAuto, rfAttenuation);
            wcdma.ConfigureDigitalEdgeTrigger("", digitalEdgeSource, digitalEdge, triggerDelay, enableTrigger);
            wcdma.ConfigureContiguousCarriers("", numberOfCarriers, carrierAtCenterFrequency);

            if (autoLevel)
            {
                wcdma.AutoLevel("", measurementInterval, out referenceLevel);
                Console.WriteLine("Reference Level (dBm)      : {0}", referenceLevel);
            }
            else
            {
                wcdma.ConfigureReferenceLevel("", referenceLevel);

            }
            wcdma.SelectMeasurements("", measurement, enableAllTraces);
            wcdma.Acp.Configuration.ConfigureMeasurementMethod("", measurementMethod);
            wcdma.Acp.Configuration.ConfigureAveraging("", averagingEnabled, averagingCount, averagingType);
            wcdma.Acp.Configuration.ConfigureSweepTime("", sweepTimeAuto, sweepTimeInterval);
            wcdma.Acp.Configuration.ConfigureNoiseCompensationEnabled("", noiseCompensationEnabled);
            wcdma.Acp.Configuration.ConfigureNumberOfOffsets("", numberOfOffsets);
            wcdma.Acp.Configuration.ConfigureOffsetPowerReference("", offsetPowerReferenceCarrier, offsetPowerReferenceSpecific);
            wcdma.Initiate("", "");
        }

        private void RetrieveResults()
        {
            wcdma.Acp.Results.FetchOffsetMeasurementArray("", timeout, ref lowerRelativePower, ref upperRelativePower,
                                                          ref lowerAbsolutePower,ref upperAbsolutePower);

            wcdma.Acp.Results.FetchCarrierMeasurementArray("", timeout, ref absolutePower, ref relativePower);

            wcdma.Acp.Results.FetchTotalCarrierPower("", timeout, out totalCarrierPower);            

            wcdma.Acp.Results.FetchSpectrum("", timeout, ref spectrum);
        }

        private void PrintResults()
        {
            Console.WriteLine("\nTotal Carrier Power  (dBm) : {0}", totalCarrierPower);

            Console.WriteLine("\nCarrier Measurements       :");
            for (int i = 0; i < absolutePower.Length; i++)
            {
                Console.WriteLine("\nCarrier                    : {0}", i);
                Console.WriteLine("Absolute Power (dBm)       : {0}", absolutePower[i]);
                Console.WriteLine("Relative Power (dB)        : {0}", relativePower[i]);
            }
            Console.WriteLine("\nOffset Channel Measurements:");
            for (int i = 0; i < lowerRelativePower.Length; i++)
            {
                Console.WriteLine("\nOffset                     : {0}", i);
                Console.WriteLine("Lower Relative Power (dB)  : {0}", lowerRelativePower[i]);
                Console.WriteLine("Upper Relative Power (dB)  : {0}", upperRelativePower[i]);
                Console.WriteLine("Lower Absolute Power (dBm) : {0}", lowerAbsolutePower[i]);
                Console.WriteLine("Upper Absolute Power (dBm) : {0}", upperAbsolutePower[i]);
            }
        }

        void CloseSession()
        {
            if (wcdma != null)
            {
                wcdma.Dispose();
                wcdma = null;
            }
            if (instrSession != null)
            {
                instrSession.Close();
                instrSession = null;
            }
        }

        static private void DisplayError(Exception ex)
        {
            Console.WriteLine("ERROR:\n" + ex.GetType() + ": " + ex.Message);
        }

    }
}

File: examples/nirfmx/wcdma/c/RFmxWcdmaAcpMultiCarrier/RFmxWcdmaAcpMultiCarrier.c
// Steps:
// 1. Open a new RFmx Session.
// 2. Configure Frequency Reference.
// 3. Configure basic signal properties (Center Frequency, RF Attenuation and External Attenuation)
// 4. Configure Trigger Type and Trigger Parameters.
// 5. Configure Contiguous Carriers.
// 6. Configure Reference Level.
// 7. Select ACP measurement and enable Traces.
// 8. Configure Measurement Method.
// 9. Configure Averaging Parameters for ACP measurement.
// 10. Configure Sweep Time Parameters.
// 11. Configure Noise Compensation Parameter.
// 12. Configure Number of offset channels. 
// 13. Configure Offset Power Reference
// 14. Initiate the Measurement.
// 15. Fetch ACP Measurements and Traces.
// 16. Close RFmx Session.

#include <stdio.h>
#include <conio.h>
#include <stdlib.h>

#include "niRFmxWCDMA.h"


/* Maximum size of an error message */
#define MAX_ERROR_DESCRIPTION       4096
#define NUMBER_OF_CARRIERS			2
#define NUMBER_OF_OFFSETS			2

int main (int argc, char *argv[])
{
    //RFSA Configuration	
    char *rfsaResourceName = "RFSA";
    niRFmxInstrHandle instrumentHandle = NULL;

    char errorMessage[MAX_ERROR_DESCRIPTION] = {0};
    int32 error = 0;
    int32 lastErrorCode = 0;
	int32 i;

    int32 enableAllTraces = RFMXWCDMA_VAL_TRUE;
    int32 measurement = RFMXWCDMA_VAL_ACP;
	char * frequencyReferenceSource =  RFMXWCDMA_VAL_ONBOARD_CLOCK_STR;
    float64 frequencyReferenceFrequency = 10e6;/*(Hz) */

    float64 centerFrequency =1.95e9;/*(Hz) */
    float64 externalAttenuation = 0.000000;/*(dB) */

	int32 RFAttenuationAuto = RFMXWCDMA_VAL_RF_ATTENUATION_AUTO_TRUE;
    float64 RFAttenuation = 10.0;/*(dB) */

    int32 enableTrigger = RFMXWCDMA_VAL_FALSE;
    char * digitalEdgeSource = RFMXWCDMA_VAL_PFI0_STR;
	int32 digitalEdge = RFMXWCDMA_VAL_DIGITAL_EDGE_RISING_EDGE;  
    float64 triggerDelay = 0.000000;/*(s) */
    int32 carrierAtCenterFrequency = -1;
   
    float64 measurementInterval = 0.010000;/*(s) */
    float64 referenceLevel = 0.000000;/*(dBm) */
    int32 measurementMethod = RFMXWCDMA_VAL_ACP_MEASUREMENT_METHOD_NORMAL;  
    int32 averagingEnabled = RFMXWCDMA_VAL_ACP_AVERAGING_ENABLED_FALSE;  
    int32 averagingCount = 10;
    int32 averagingType = RFMXWCDMA_VAL_ACP_AVERAGING_TYPE_RMS;  

	int32 autoLevel = RFMXWCDMA_VAL_TRUE;

    int32 sweepTimeAuto = RFMXWCDMA_VAL_ACP_SWEEP_TIME_AUTO_TRUE;  
    float64 sweepTimeInterval = 0.000667;/*(s) */

    int32 noiseCompensationEnabled = RFMXWCDMA_VAL_ACP_NOISE_COMPENSATION_ENABLED_FALSE;  
    int32 offsetPowerReferenceCarrier = RFMXWCDMA_VAL_ACP_OFFSET_POWER_REFERENCE_CARRIER_COMPOSITE;
    int32 offsetPowerReferenceSpecific = 0;

	float64 timeout = 10.000000;/*(s) */

	float64 absolutePower[NUMBER_OF_CARRIERS] = {0};			/*(dBm) */
	float64 relativePower[NUMBER_OF_CARRIERS] = {0};			/*(dB) */
    float64 lowerAbsolutePower[NUMBER_OF_OFFSETS] = {0};		/*(dBm) */ 
    float64 upperAbsolutePower[NUMBER_OF_OFFSETS] = {0};		/*(dBm) */    
    float64 lowerRelativePower[NUMBER_OF_OFFSETS] = {0};		/*(dB) */ 
    float64 upperRelativePower[NUMBER_OF_OFFSETS] = {0};		/*(dB) */ 

    float64 totalCarrierPower = 0.000000;/*(dBm) */

	int32 actualArraySize = 0;
	float64 x0=0.0,dx=0.0;
    float32 *spectrum = NULL;/*(dBm) */


    /* Initialize a session */
    RFmxCheckWarn(RFmxWCDMA_Initialize(rfsaResourceName, "", &instrumentHandle, NULL));

    RFmxCheckWarn(RFmxWCDMA_CfgFrequencyReference(instrumentHandle, "",frequencyReferenceSource, frequencyReferenceFrequency));
    RFmxCheckWarn(RFmxWCDMA_CfgFrequency(instrumentHandle, "", centerFrequency));
    RFmxCheckWarn(RFmxWCDMA_CfgExternalAttenuation(instrumentHandle, "", externalAttenuation));
    RFmxCheckWarn(RFmxWCDMA_CfgRFAttenuation(instrumentHandle, "", RFAttenuationAuto, RFAttenuation));
    RFmxCheckWarn(RFmxWCDMA_CfgDigitalEdgeTrigger(instrumentHandle, "", digitalEdgeSource, digitalEdge, triggerDelay, enableTrigger));
    RFmxCheckWarn(RFmxWCDMA_CfgContiguousCarriers(instrumentHandle, "", NUMBER_OF_CARRIERS, carrierAtCenterFrequency));
    if( autoLevel )
    {
        RFmxCheckWarn(RFmxWCDMA_AutoLevel(instrumentHandle, "", measurementInterval, &referenceLevel));		
        printf("Reference level (dBm)           : %f\n", referenceLevel);
    }
    else 
	{
        RFmxCheckWarn(RFmxWCDMA_CfgReferenceLevel(instrumentHandle, "", referenceLevel));
    }
    RFmxCheckWarn(RFmxWCDMA_SelectMeasurements(instrumentHandle, "", measurement, enableAllTraces));
    RFmxCheckWarn(RFmxWCDMA_ACPCfgMeasurementMethod(instrumentHandle, "", measurementMethod));
    RFmxCheckWarn(RFmxWCDMA_ACPCfgAveraging(instrumentHandle, "", averagingEnabled, averagingCount, averagingType));
    RFmxCheckWarn(RFmxWCDMA_ACPCfgSweepTime(instrumentHandle, "", sweepTimeAuto, sweepTimeInterval));
    RFmxCheckWarn(RFmxWCDMA_ACPCfgNoiseCompensationEnabled(instrumentHandle, "", noiseCompensationEnabled));
    RFmxCheckWarn(RFmxWCDMA_ACPCfgNumberOfOffsets(instrumentHandle, "", NUMBER_OF_OFFSETS));
    RFmxCheckWarn(RFmxWCDMA_ACPCfgOffsetPowerReference(instrumentHandle, "", offsetPowerReferenceCarrier, offsetPowerReferenceSpecific));
    RFmxCheckWarn(RFmxWCDMA_Initiate(instrumentHandle, "", ""));

	/* Fetch the offsetmeasurements array */
	RFmxCheckWarn(RFmxWCDMA_ACPFetchOffsetMeasurementArray(instrumentHandle, "", timeout, lowerRelativePower, upperRelativePower, lowerAbsolutePower, upperAbsolutePower, NUMBER_OF_OFFSETS, NULL));
	
	/* Fetch the carrier measurements array */
	RFmxCheckWarn(RFmxWCDMA_ACPFetchCarrierMeasurementArray(instrumentHandle, "", timeout,absolutePower, relativePower,NUMBER_OF_CARRIERS, NULL));
	
	RFmxCheckWarn(RFmxWCDMA_ACPFetchTotalCarrierPower(instrumentHandle, "", timeout, &totalCarrierPower));
    
	RFmxCheckWarn(RFmxWCDMA_ACPFetchSpectrum(instrumentHandle, "", timeout, NULL, NULL, NULL, 0, 
										   &actualArraySize));
	if( actualArraySize > 0 )
	{
		spectrum = (float32 *)malloc(sizeof(float32) * actualArraySize);
		if(spectrum)
		{
				RFmxCheckWarn(RFmxWCDMA_ACPFetchSpectrum(instrumentHandle, "", timeout, 
					&x0, &dx, spectrum, actualArraySize, NULL));	
		}
		else
		{
			printf("malloc failed.\n");
			goto Error;
		}
	}
    
    printf("\nTotal Aggregated Power (dBm)    : %f\n", totalCarrierPower);

	printf("\nCarrier Measurements            : \n");
	for(i=0;i<NUMBER_OF_CARRIERS;i++)
	{
		printf("Carrier                         : %d\n", i);
        printf("Absolute Power (dBm)            : %f\n", absolutePower[i]);
        printf("Relative Power (dB)             : %f\n", relativePower[i]);
		printf("-------------------------------------------------\n");
	}
	
	printf("\nOffset Channel Measurements     : \n");
	for(i=0;i<NUMBER_OF_OFFSETS;i++)
	{
		printf("Offset                          : %d\n", i);
        printf("Lower Relative Power (dB)       : %f\n", lowerRelativePower[i]);
        printf("Upper Relative Power (dB)       : %f\n", upperRelativePower[i]);
        printf("Lower Absolute Power (dBm)      : %f\n", lowerAbsolutePower[i]);
        printf("Upper Absolute Power (dBm)      : %f\n", upperAbsolutePower[i]);
		printf("-------------------------------------------------\n");
	}


Error:
    if( error )
    {
        RFmxWCDMA_GetError(instrumentHandle, &lastErrorCode, MAX_ERROR_DESCRIPTION, errorMessage);
        if (error < 0)
            printf("ERROR: %s\n", errorMessage);
        else
            printf("WARNING: %s\n", errorMessage);
    }
    if(instrumentHandle)
    {
        RFmxWCDMA_Close(instrumentHandle, RFMXWCDMA_VAL_FALSE);
    }

    if (spectrum)
    {
        free(spectrum);
    }
  
    printf("Press any key to exit\n");
    _getch();

    return error;
}
